Modules/arraymodule.c

```diff
@@ -1115,18 +1115,39 @@ array_array_count(arrayobject *self, PyObject *v)
 array.array.index
 
     v: object
+    start: slice_index(accept={int}) = 0
+    stop: slice_index(accept={int}, c_default="PY_SSIZE_T_MAX") = sys.maxsize
     /
 
 Return index of first occurrence of v in the array.
+
+Raises ValueError if the value is not present.
 [clinic start generated code]*/
 
 static PyObject *
-array_array_index(arrayobject *self, PyObject *v)
-/*[clinic end generated code: output=d48498d325602167 input=cf619898c6649d08]*/
+array_array_index_impl(arrayobject *self, PyObject *v, Py_ssize_t start,
+                       Py_ssize_t stop)
+/*[clinic end generated code: output=c45e777880c99f52 input=6efdb3dae558556c]*/
 {
-    Py_ssize_t i;
+    Py_ssize_t i, len;
 
-    for (i = 0; i < Py_SIZE(self); i++) {
+    len = Py_SIZE(self);
+
+    if (start < 0) {
+        start += len;
+
+        if (start < 0) {
+            start = 0;
+        }
+    }
+
+    if (stop < 0) {
+        stop += len;
+    } else if (stop > len) {
+        stop = len;
+    }
+
+    for (i = start; i < stop; i++) {
```

Comment:
An arbitrary Python code can be executed in `PyObject_RichCompareBool()`. It can change the size of the array. As result, the index can become out of the array limits.

Use the same loop condition as in list.index to prevent this.

Reply:
Yes, I was able to reproduce this, using an object that contained a reference to the array, which in its `__eq__` mutates the array, causing an assertion failure. At this point in time, my code has boiled down to an almost identical copy of `list_index`, except I do not see the reason for truncating the stop index, as is done there.